File: apps/api/internal/assistant/intent_llm.go
package assistant

import (
	"context"
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/Ciao1019/Petrichor/apps/api/ent"
	"github.com/Ciao1019/Petrichor/apps/api/internal/ai"
)

const (
	intentLLMConfidenceThreshold = 0.5
	intentLLMTimeout             = 5 * time.Second
)

type intentLLMObject struct {
	Domains    []string `json:"domains"`
	Confidence float64  `json:"confidence"`
	Rationale  string   `json:"rationale"`
}

// RouteAssistantIntentWithLLM 规则优先；低置信度或 admin 候选时用轻量 JSON 分类覆盖。
// LLM 失败时静默回退规则结果，不阻断对话。
func RouteAssistantIntentWithLLM(
	ctx context.Context,
	gen *ai.Generation,
	modelCfg *ent.AIModelConfig,
	userText string,
	focus *AssistantFocus,
	recentToolNames []string,
	recentIntentDomains []AgentDomainId,
	rulesRoute *IntentRouteResult,
) IntentRouteResult {
	rules := IntentRouteResult{}
	if rulesRoute != nil {
		rules = *rulesRoute
	} else {
		rules = RouteAssistantIntent(userText, focus, recentToolNames, recentIntentDomains)
	}
	finalize := func(route IntentRouteResult) IntentRouteResult {
		route.Domains = withStickyAdminDomain(route.Domains, recentIntentDomains)
		if route.Source == "" {
			route.Source = "rules"
		}
		return route
	}
	rulesResult := finalize(IntentRouteResult{
		Domains:    rules.Domains,
		Confidence: rules.Confidence,
		Rationale:  rules.Rationale,
		Source:     "rules",
	})
	if !NeedsIntentLLM(rules) || gen == nil || modelCfg == nil {
		return rulesResult
	}
	rulesHadAdmin := hasAdminDomainCandidate(rules.Domains)

	llmCtx, cancel := context.WithTimeout(ctx, intentLLMTimeout)
	defer cancel()
	llm, err := classifyIntentWithLLM(llmCtx, gen, modelCfg, userText, focus, recentToolNames, recentIntentDomains, rules)
	if err != nil {
		if rulesHadAdmin {
			rulesResult.Rationale = rules.Rationale + ";admin-domain-kept-on-llm-failure"
			return finalize(rulesResult)
		}
		return finalize(rulesResult)
	}
	merged := mergeAdminDomainFromRules(llm.Domains, rules.Domains)
	if merged.kept {
		llm.Rationale = llm.Rationale + ";admin-domain-kept-from-rules"
	}
	llm.Domains = merged.domains
	return finalize(llm)
}

func classifyIntentWithLLM(
	ctx context.Context,
	gen *ai.Generation,
	modelCfg *ent.AIModelConfig,
	userText string,
	focus *AssistantFocus,
	recentToolNames []string,
	recentIntentDomains []AgentDomainId,
	rules IntentRouteResult,
) (IntentRouteResult, error) {
	system := strings.Join([]string{
		"你是 Petrichor 站内助手的意图分类器。",
		"只输出 JSON，选择本轮需要装载的工具域。",
		`格式：{"domains":["knowledge"],"confidence":0.8,"rationale":"短句"}`,
		"可选域：knowledge、system、content_write、admin。",
		"domains 数量 1~2。",
		"用户明确要求改配置/密钥/公开问答开关时必须含 admin。",
		"上一轮意图域已含 admin 时，除非用户明确改聊无关只读话题，否则继续包含 admin。",
		"知识库问答通常同时含 system。",
		"rationale 用中文短句，不超过 80 字。",
		"不要编造域。",
	}, "\n")
	prompt := strings.Join([]string{
		fmt.Sprintf("用户消息：%s", emptyHint(userText)),
		fmt.Sprintf("当前 focus：%s", summarizeFocus(focus)),
		fmt.Sprintf("近期工具：%s", joinOrNone(recentToolNames)),
		fmt.Sprintf("上一轮意图域：%s", joinDomainsOrNone(recentIntentDomains)),
		fmt.Sprintf("规则初筛 domains：%s", joinDomainsOrNone(rules.Domains)),
		fmt.Sprintf("规则置信度：%.2f", rules.Confidence),
		fmt.Sprintf("规则 rationale：%s", rules.Rationale),
	}, "\n")

	result, err := gen.Chat(ctx, modelCfg, []ai.ChatMessage{
		{Role: "system", Content: system},
		{Role: "user", Content: prompt},
	})
	if err != nil {
		return IntentRouteResult{}, err
	}
	obj, err := parseIntentLLMJSON(result.Content)
	if err != nil {
		return IntentRouteResult{}, err
	}
	domains := make([]AgentDomainId, 0, maxPrimaryIntentDomains)
	seen := map[AgentDomainId]struct{}{}
	for _, raw := range obj.Domains {
		d := AgentDomainId(strings.TrimSpace(raw))
		if !isValidDomain(d) {
			continue
		}
		if _, ok := seen[d]; ok {
			continue
		}
		seen[d] = struct{}{}
		domains = append(domains, d)
		if len(domains) >= maxPrimaryIntentDomains {
			break
		}
	}
	if len(domains) == 0 {
		return IntentRouteResult{}, fmt.Errorf("意图 LLM 未返回有效 domains")
	}
	conf := obj.Confidence
	if conf < 0 {
		conf = 0
	}
	if conf > 1 {
		conf = 1
	}
	rationale := strings.TrimSpace(obj.Rationale)
	if rationale == "" {
		rationale = "llm"
	}
	if len([]rune(rationale)) > 80 {
		rationale = string([]rune(rationale)[:80])
	}
	return IntentRouteResult{
		Domains:    withAuxiliaryDomains(domains),
		Confidence: conf,
		Rationale:  rationale,
		Source:     "llm",
	}, nil
}

func parseIntentLLMJSON(raw string) (intentLLMObject, error) {
	text := strings.TrimSpace(raw)
	text = stripJSONFence(text)
	if obj, ok := tryUnmarshalIntent(text); ok {
		return obj, nil
	}
	re := regexp.MustCompile(`\{[\s\S]*\}`)
	loc := re.FindString(text)
	if loc == "" {
		return intentLLMObject{}, fmt.Errorf("未找到 JSON 对象")
	}
	if obj, ok := tryUnmarshalIntent(loc); ok {
		return obj, nil
	}
	return intentLLMObject{}, fmt.Errorf("意图 JSON 解析失败")
}

func tryUnmarshalIntent(raw string) (intentLLMObject, bool) {
	var obj intentLLMObject
	if err := json.Unmarshal([]byte(raw), &obj); err != nil {
		return intentLLMObject{}, false
	}
	return obj, true
}

func stripJSONFence(s string) string {
	s = strings.TrimSpace(s)
	if strings.HasPrefix(s, "```") {
		s = strings.TrimPrefix(s, "```json")
		s = strings.TrimPrefix(s, "```JSON")
		s = strings.TrimPrefix(s, "```")
		if idx := strings.LastIndex(s, "```"); idx >= 0 {
			s = s[:idx]
		}
	}
	return strings.TrimSpace(s)
}

type mergeAdminResult struct {
	domains []AgentDomainId
	kept    bool
}

func mergeAdminDomainFromRules(llmDomains, rulesDomains []AgentDomainId) mergeAdminResult {
	if !hasAdminDomainCandidate(rulesDomains) {
		return mergeAdminResult{domains: llmDomains}
	}
	if hasAdminDomainCandidate(llmDomains) {
		return mergeAdminResult{domains: llmDomains}
	}
	return mergeAdminResult{domains: withAuxiliaryDomains(append(append([]AgentDomainId{}, llmDomains...), DomainAdmin)), kept: true}
}

func isValidDomain(d AgentDomainId) bool {
	switch d {
	case DomainKnowledge, DomainSystem, DomainContentWrite, DomainAdmin:
		return true
	default:
		return false
	}
}

func summarizeFocus(focus *AssistantFocus) string {
	if focus == nil {
		return "无"
	}
	parts := make([]string, 0, 4)
	if focus.KnowledgeBaseID != nil {
		parts = append(parts, "knowledgeBaseId="+*focus.KnowledgeBaseID)
	}
	if focus.ArticleID != nil {
		parts = append(parts, "articleId="+*focus.ArticleID)
	}
	if focus.DocumentID != nil {
		parts = append(parts, "documentId="+*focus.DocumentID)
	}
	if len(parts) == 0 {
		return "无"
	}
	return strings.Join(parts, ", ")
}

func emptyHint(s string) string {
	if strings.TrimSpace(s) == "" {
		return "（空）"
	}
	return s
}

func joinOrNone(items []string) string {
	if len(items) == 0 {
		return "无"
	}
	limit := 8
	if len(items) < limit {
		limit = len(items)
	}
	return strings.Join(items[:limit], ", ")
}

func joinDomainsOrNone(domains []AgentDomainId) string {
	if len(domains) == 0 {
		return "无"
	}
	parts := make([]string, len(domains))
	for i, d := range domains {
		parts[i] = string(d)
	}
	return strings.Join(parts, ", ")
}

File: apps/web/src/server/assistant/intent-llm.ts
import { generateObject, type LanguageModel } from "ai"
import { z } from "zod"
import { createLogger, toLogError } from "@/lib/logger"
import type { AgentDomainId, AssistantFocus, IntentRouteResult, IntentRouteSource } from "./domain-types"
import {
    routeAssistantIntent,
    withAuxiliaryDomains,
    MAX_PRIMARY_INTENT_DOMAINS,
    hasAdminDomainCandidate,
    withStickyAdminDomain,
    mergeAdminDomainFromRules,
} from "./intent-router"

const log = createLogger("assistant-intent-llm")

export const INTENT_LLM_CONFIDENCE_THRESHOLD = 0.5
export const INTENT_LLM_TIMEOUT_MS = 5_000
export const INTENT_ROUTE_PART_TYPE = "data-intent-route"

const AGENT_DOMAIN_IDS = [
    "knowledge",
    "doc_library",
    "system",
    "content_write",
    "admin",
] as const satisfies readonly AgentDomainId[]

export const DOMAIN_LABELS: Record<AgentDomainId, string> = {
    knowledge: "知识库",
    doc_library: "文档库",
    system: "系统",
    content_write: "内容写入",
    admin: "管理",
}

const intentLlmSchema = z.object({
    domains: z.array(z.enum(AGENT_DOMAIN_IDS)).min(1).max(MAX_PRIMARY_INTENT_DOMAINS),
    confidence: z.number().min(0).max(1),
    rationale: z.string().trim().min(1).max(80),
})

export type IntentRouteUiResult = IntentRouteResult & {
    source: IntentRouteSource
}

export type IntentRoutePartData = {
    status: "running" | "done"
    source?: IntentRouteSource
    domains?: AgentDomainId[]
    confidence?: number
    rationale?: string
    label: string
}

/**
 * 仅低置信度或规则命中 admin 时调用意图 LLM。
 * content_write 已会话常驻，不再为写域强制复核。
 */
export function needsIntentLlm(route: IntentRouteResult): boolean {
    return route.confidence < INTENT_LLM_CONFIDENCE_THRESHOLD || hasAdminDomainCandidate(route.domains)
}

export function formatIntentRouteLabel(input: {
    domains: AgentDomainId[]
    source: IntentRouteSource
    confidence: number
}): string {
    const domainText = input.domains.map((domain) => DOMAIN_LABELS[domain] ?? domain).join(" · ")
    const sourceText = input.source === "llm" ? "智能识别" : "规则"
    const confidenceText = `${Math.round(input.confidence * 100)}%`
    return `${domainText}（${sourceText} · ${confidenceText}）`
}

export function toIntentRoutePartData(
    route: IntentRouteUiResult,
    status: "running" | "done" = "done",
): IntentRoutePartData {
    if (status === "running") {
        return { status: "running", label: "正在识别意图…" }
    }
    return {
        status: "done",
        source: route.source,
        domains: route.domains,
        confidence: route.confidence,
        rationale: route.rationale,
        label: formatIntentRouteLabel(route),
    }
}

export function domainsEqual(a: AgentDomainId[], b: AgentDomainId[]): boolean {
    if (a.length !== b.length) return false
    return a.every((domain, index) => domain === b[index])
}

/** 同一条助手消息里只保留最后一条 intent-route（防御 id 合并失败或历史脏数据） */
export function dedupeIntentRouteParts(parts: unknown): unknown[] {
    if (!Array.isArray(parts)) return []
    let lastIntentIndex = -1
    for (let index = 0; index < parts.length; index += 1) {
        const part = parts[index]
        if (!part || typeof part !== "object") continue
        if ((part as { type?: unknown }).type === INTENT_ROUTE_PART_TYPE) {
            lastIntentIndex = index
        }
    }
    if (lastIntentIndex < 0) return parts
    return parts.filter((part, index) => {
        if (!part || typeof part !== "object") return true
        if ((part as { type?: unknown }).type !== INTENT_ROUTE_PART_TYPE) return true
        return index === lastIntentIndex
    })
}

/**
 * 规则优先；低置信度或 admin 候选时用轻量 generateObject 覆盖。
 * LLM 失败 / 超时 / abort 时静默回退规则结果，不阻断对话。
 */
export async function routeAssistantIntentWithLlm(input: {
    userText: string
    focus: AssistantFocus | null
    recentToolNames: string[]
    recentIntentDomains?: AgentDomainId[]
    model: LanguageModel
    signal?: AbortSignal
    /** 若调用方已跑过规则，传入可避免重复打分 */
    rulesRoute?: IntentRouteResult
}): Promise<IntentRouteUiResult> {
    const rules = input.rulesRoute ?? await routeAssistantIntent({
        userText: input.userText,
        focus: input.focus,
        recentToolNames: input.recentToolNames,
        recentIntentDomains: input.recentIntentDomains,
    })
    const finalize = (route: IntentRouteUiResult): IntentRouteUiResult => ({
        ...route,
        domains: withStickyAdminDomain(route.domains, input.recentIntentDomains),
    })
    const rulesResult = finalize({ ...rules, source: "rules" })

    if (!needsIntentLlm(rules)) {
        return rulesResult
    }

    const rulesHadAdmin = hasAdminDomainCandidate(rules.domains)

    try {
        const llm = await classifyIntentWithLlm({
            userText: input.userText,
            focus: input.focus,
            recentToolNames: input.recentToolNames,
            recentIntentDomains: input.recentIntentDomains,
            rules,
            model: input.model,
            signal: input.signal,
        })
        // 规则已命中 admin 时不允许 LLM 剥掉；再叠加上一轮 admin 粘性
        const merged = mergeAdminDomainFromRules(llm.domains, rules.domains)
        return finalize({
            ...llm,
            domains: merged.domains,
            rationale: merged.kept
                ? `${llm.rationale};admin-domain-kept-from-rules`
                : llm.rationale,
        })
    } catch (error) {
        log.warn({
            err: toLogError(error),
            rulesHadAdmin,
        }, "意图模型分类失败，回退到规则")
        if (rulesHadAdmin) {
            return finalize({
                ...rulesResult,
                rationale: `${rules.rationale ?? "rules"};admin-domain-kept-on-llm-failure`,
            })
        }
        return finalize(rulesResult)
    }
}

export async function classifyIntentWithLlm(input: {
    userText: string
    focus: AssistantFocus | null
    recentToolNames: string[]
    recentIntentDomains?: AgentDomainId[]
    rules: IntentRouteResult
    model: LanguageModel
    signal?: AbortSignal
}): Promise<IntentRouteUiResult> {
    const focusHint = summarizeFocus(input.focus)
    const recentHint = input.recentToolNames.length > 0
        ? input.recentToolNames.slice(0, 8).join(", ")
        : "无"
    const recentDomainsHint = input.recentIntentDomains && input.recentIntentDomains.length > 0
        ? input.recentIntentDomains.join(", ")
        : "无"

    const timeout = AbortSignal.timeout(INTENT_LLM_TIMEOUT_MS)
    const abortSignal = input.signal
        ? AbortSignal.any([input.signal, timeout])
        : timeout

    const result = await generateObject({
        model: input.model,
        schema: intentLlmSchema,
        temperature: 0,
        abortSignal,
        system: [
            "你是 Petrichor 站内助手的意图分类器。",
            "只输出 JSON，选择本轮需要装载的工具域。",
            "可选域：knowledge（知识库）、doc_library（文档库）、system（系统概览/进度/引用）、content_write（写/改/移动/迁移文章、分享）、admin（模型配置/API Key/公开问答）。",
            "规则：用户明确要求改配置/密钥/公开问答开关时必须含 admin。",
            "规则：上一轮意图域已含 admin 时，本轮除非用户明确改聊无关只读话题，否则继续包含 admin。",
            "规则：知识库或文档库问答通常同时含 system（便于引用与进度）。",
            "说明：content_write 工具已会话常驻，芯片可标可不标；纯问答不必强行挂 content_write。",
            "rationale 用中文短句说明理由，不超过 80 字。",
            "不要编造域。",
        ].join("\n"),
        prompt: [
            `用户消息：${input.userText || "（空）"}`,
            `当前 focus：${focusHint}`,
            `近期工具：${recentHint}`,
            `上一轮意图域：${recentDomainsHint}`,
            `规则初筛 domains：${input.rules.domains.join(", ")}`,
            `规则置信度：${input.rules.confidence}`,
            `规则 rationale：${input.rules.rationale ?? ""}`,
        ].join("\n"),
    })

    const domains = withAuxiliaryDomains(result.object.domains.slice(0, MAX_PRIMARY_INTENT_DOMAINS))
    return {
        domains,
        confidence: result.object.confidence,
        rationale: result.object.rationale.trim(),
        source: "llm",
    }
}

function summarizeFocus(focus: AssistantFocus | null): string {
    if (!focus) return "无"
    const parts: string[] = []
    if (focus.knowledgeBaseId != null) parts.push(`knowledgeBaseId=${focus.knowledgeBaseId}`)
    if (focus.articleId != null) parts.push(`articleId=${focus.articleId}`)
    if (focus.libraryId != null) parts.push(`libraryId=${focus.libraryId}`)
    if (focus.documentId != null) parts.push(`documentId=${focus.documentId}`)
    return parts.length > 0 ? parts.join(", ") : "无"
}
